### ISSUE TITLE
Question: selected 'centered' value reverts to 'thing' on add-on restart

### ISSUE THREAD

**Issue by flatsiedatsie** [WebThingsIO/gateway-addon-python#48]
I have a simple MySensors relay device which is given an OnOff capacity. After turning it into a thing I select that on-off property to be centered:

![anemone-on](https://user-images.githubusercontent.com/805405/59146841-a972d480-89e2-11e9-92d9-ae5fb87a4c4a.png)

When the MySensors add-on is restarted, it reverts back to the basic 'thing' display:

![anemone-thing](https://user-images.githubusercontent.com/805405/59146848-c5767600-89e2-11e9-96f5-8f8fbac4598f.png)

I'm trying to understand the cause. My first suspicion is this:
- On restart of the add-on, it asks all mysensors devices to re-present themselves. Their response to this discovery request is what generates the devices.
- As the device presents its name, it's recreated as a device by the adapter. Then `self.handle_device_added()` is called.
- As each property arrives, it is created as well. Then `self.notify_property_changed()` is called, followed by `self.handle_device_added()`.

Could it be that calling `self.handle_device_added()` after the name of the thing arrives, but before any properties are added, is creating an 'empty' version of thing, and this resets which property gets centered?

Also: is calling `self.handle_device_added()` after adding properties necessary?

**Comment by flatsiedatsie** [comment]
Diving into the code a bit, I was wondering if I understand this correctly:
- `self.handle_device_added()` sends the entire device object to the gateway, including all the properties. Properties can be added to a device object even though `self.handle_device_added()` hasn't been called yet.
- Initializing properties on the gateway (telling the Gateway they exist) should be done by the above method.
- `self.notify_property_changed()` should/cannot be called before `self.handle_device_added()` has been done. It's only supposed to be used when a property has already been reported to exist to the gateway, and its value has changed. It's not the _property_ equal of _handle_device_added_. It does not function as "handle_property_added".

If this understanding is correct, perhaps this would be best?
1. Create a device object and one by one add all properties to it.
2. Once done, call `handle_device_added()`.
3. If the entire device has been presented (including properties) and a property value changes, only then call `notify_property_changed()`.

Is this the 'correct' / intended way of doing it?

If so, the problem is that I don't know when all properties have presented themselves. The MySensors nodes don't communicate how many properties they have in total. So I don't know when the device is 'complete' and I can finally call `handle_device_added`.

So to summarize:
- Is my understanding of the 'intended' way of doing things now correct?
- Could this be the cause of the Gateway 'forgetting' which property to put in the center bubble?
If so, I will change how the add-on functions. But then the question becomes:
- How could I deal with not knowing when the device object is 'complete'?

P.S. Currently the add-on sends a `handle_device_added()` message to the gateway everytime a property is added to a device. Could this be the cause of the [MaxListenersExceededWarning](https://github.com/mozilla-iot/gateway/issues/1128)? Is the add-on flooding the Gateway?

**Comment by dhylands** [comment]
What you're saying sounds about right to me. The zgbee and zwave adapters often don't hear from all of the devices at startup time (battery powered devices sleep and may only report in once every few hours), so they collect the property information at pairing time, and then cache (persist) that information so that its available in the future even when the devices are asleep.

**Comment by dhylands** [comment]
I've also noticed the behaviour you've seen where the "selectedCapability" seems to get forgotten or lost. The adapters actually don't know what the selected capability is, so this is probably an issue on the gateway side of things.

**Comment by mrstegeman** [comment]
@flatsiedatsie You’re on the right track. The cause of the empty bubble is that `self._type` (the `@type` array in the thing description) is not being populated the same way in your `Device` class the second time it’s created, or at least not at the point that the gateway first receives it.

The best thing to do is to wait until you’re reasonably certain you’ve fully populated the device and its properties before calling `handle_device_added()`.

**Comment by flatsiedatsie** [comment]
Thank you both! I will rewrite the add-on to always enable persistence, and to first completely fill the device object before calling `handle_device_added()`. When the device first becomes available it will still do it the old way (no choice there).

The reason I personally kept persistence turned off is that I like knowing which devices have actually responded to the add-on's discovery request. Devices which have not responded would stay in the 'disconnected' state.

Going through the code I think I saw some code that would allow the add-on to set the connected and disconnected state, correct? Then I could populate the devices via persistence, but also set them as 'disconnected' initially until they actually respond.

Is that possible?

**Comment by flatsiedatsie** [comment]
For now the code just sets the thing to disconnected immediately after presenting it to the WebThings Gateway.

As suspected, the issue was with the add-on's way of doing things. I am closing the issue.

**Comment by mrstegeman** [comment]
If you didn’t already find it, you can use `connected_notify()`.

https://github.com/mozilla-iot/gateway-addon-python/blob/master/gateway_addon/device.py#L180

**Comment by flatsiedatsie** [comment]
I found it :-)
